src/main/scala/com/databricks/spark/avro/AvroRelation.scala

```diff
@@ -40,8 +46,40 @@ case class AvroRelation(location: String)(@transient val sqlContext: SQLContext)
     convertedSchema
   }
 
+  // There are several properties of Avro that we have to account for in this method:
+  // 1) Avro uses Utf8 strings, so we want to convert them to java.lang.String for SparkSQL,
+  //    including making the appropriate recursive calls.
+  // 2) Avro map keys are always strings, so we don't need to recurse on them when processing maps.
+  // 3) Byte arrays and ByteBuffers are reused by Avro, so we must copy them out.
+  def convertToSparkSQL(obj: Any): Any = {
+    obj match {
+      case u: Utf8 => u.toString
+      case m: Map[Any, Any] => m.map(x => (x._1.toString, convertToSparkSQL(x._2)))
+      case a: GenericData.Array[Any] => {
+        val ar = new Array[Any](a.size)
+        var idx = 0
+        while (idx < a.size) {
+          ar(idx) = convertToSparkSQL(a(0))
+          idx += 1
+        }
+        ar.toSeq
+      }
+      case f: Fixed => f.bytes.clone
+      case e: EnumSymbol => e.toString
+      case r: Record => Row.fromSeq((0 until r.getSchema.getFields.size).map { i =>
+        convertToSparkSQL(r.get(i))
+      })
+      case h: ByteBuffer => {
+        val ar = new Array[Byte](h.remaining)
+        h.get(ar)
+        ar
+      }
+      case other => other
```

Comment:
Apparently you cannot use AnyVal in match statements. I can make a set with primitive types and check if other's type belongs there though (or there is a shorter but uglier solution).

However, it seems to me that this is unnecessary. This function is only called for whatever we parsed from avro file. If there was invalid type in the avro file, we would get an appropriate exception when reading it (I checked). As a result, this function will only get legal avro types. We are handling all of them, so no problem should occur. If there was was a union(A,B,C), the obj parameter that this function gets would be one of A,B,C, rather some class that represents unions. If the union wasn't supported by us, we would throw an exception during schema parsing, which happens before.

So I think we should't be adding this. If you still think otherwise, I'll add it.